src/eos-updater-apply.c

```diff
@@ -73,6 +74,431 @@ get_test_osname (void)
+    return NULL;
+
+  /* Filter the flatpaks this commit wants by removing any flatpaks
+   * which were already explicitly removed earlier, or which are already
+   * installed */
+  for (; i < flatpaks_this_commit_wants->len; ++i)
+    {
+      FlatpakToInstall *flatpak_to_install = g_ptr_array_index (flatpaks_this_commit_wants, i);
+
+      if (!g_hash_table_contains (explicitly_removed_flatpaks,
+                                  flatpak_to_install->app_identifier) &&
+          !g_hash_table_contains (already_installed_flatpaks,
+                                  flatpak_to_install->app_identifier))
+        {
+          g_ptr_array_add (resolved_flatpaks,
+                           flatpak_to_install_copy (flatpak_to_install));
+        }
+    }
+
+  return g_steal_pointer (&resolved_flatpaks);
+}
+
+static FlatpakInstallation *
+get_flatpak_installation (GCancellable *cancellable, GError **error)
+{
+  const gchar *override_path = g_getenv ("EOS_UPDATER_TEST_FLATPAK_INSTALLATION_DIR");
+
+  if (override_path)
+    {
+      g_autoptr(GFile) override = g_file_new_for_path (override_path);
+      return flatpak_installation_new_for_path (override, TRUE, cancellable, error);
+    }
+
+  return flatpak_installation_new_system (cancellable, error);
+}
+
+static gboolean
+pull_flatpaks (GPtrArray    *flatpaks_to_deploy,
+               GCancellable *cancellable,
+               GError       **error)
+{
+  g_autoptr(FlatpakInstallation) installation = get_flatpak_installation (cancellable,
+                                                                          error);
+  gsize i = 0;
+
+  if (!installation)
+    return FALSE;
+
+  for (; i < flatpaks_to_deploy->len; ++i)
+    {
+      FlatpakToInstall *to_install = g_ptr_array_index (flatpaks_to_deploy, i);
+      g_autoptr(GError) local_error = NULL;
+
+      /* We have to pass in a local_error instance here and check to see
+       * if it was FLATPAK_ERROR_NOT_INSTALLED - this is what will be
+       * thrown if we succeeded at pulling the flatpak into the local
+       * repository but did not deploy it (since no FlatpakInstalledRef
+       * will be returned). */
+      flatpak_installation_install_full (installation,
+                                         FLATPAK_INSTALL_FLAGS_NO_DEPLOY,
+                                         to_install->repo,
+                                         to_install->ref_kind,
+                                         to_install->app_identifier,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         cancellable,
+                                         &local_error);
+
+      /* This only happens if the wrong version of flatpak is deployed
+       * with eos-updater */
+      if (!local_error)
```

Comment:
Nitpick: eos-updater uses explicitly `NULL`. Comparisons, as the types are clearer if you use `if (local_error == NULL)` rather than `if (!local_error)`. Please try and do so throughout the patch. :-)